pkg/syncer/endpoints/endpoint_downstream_controller.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package endpoints
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	ddsif "github.com/kcp-dev/kcp/pkg/informer"
+	"github.com/kcp-dev/kcp/pkg/logging"
+)
+
+const (
+	ControllerName = "syncer-endpoint-controller"
+)
+
+func NewEndpointController(
+	syncerLogger logr.Logger,
+	downstreamClient dynamic.Interface,
+	ddsifForDownstream *ddsif.GenericDiscoveringDynamicSharedInformerFactory[cache.SharedIndexInformer, cache.GenericLister, informers.GenericInformer],
+	syncedInformers map[schema.GroupVersionResource]cache.SharedIndexInformer,
+) (*EndpointController, error) {
+	endpointsGVR := corev1.SchemeGroupVersion.WithResource("endpoints")
+
+	c := &EndpointController{
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), ControllerName),
+	}
+
+	logger := logging.WithReconciler(syncerLogger, ControllerName)
+
+	endpointsInformer, ok := syncedInformers[endpointsGVR]
+	if !ok {
+		return nil, errors.New("endpoints informer should be available")
+	}
+
+	endpointsInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.AddToQueue(obj, logger)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			c.AddToQueue(newObj, logger)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.AddToQueue(obj, logger)
+		},
+	})
+
+	return c, nil
+}
+
+type EndpointController struct {
```

Comment:
What is an endpoint controller?

Reply:
Sorry, still need to add GoDoc. Will do it.
This is part of the Proof PR (and would be finally split in a next PR when Upsyncing is finished and merged).

It is a controller that would annotate some Endpoints related to Synced services, to be UpSynced, so that a KNative controller (running against KCP) could see the Endpoint and know that the Service is effective.

In fact it is just a minimal example of GVR-specific controller. A similar one would exist to annotate (for upsyncing) Pods related to a synced deployment.

Reply:
GoDoc added